pkg/interface/src/views/apps/term/app.tsx

```diff
@@ -1,100 +1,458 @@
+      available = true;
+    }
+
+    slog.onmessage = e => {
+      onSlog(e.data);
+    }
+
+    slog.onerror = e => {
+      console.error('slog: eventsource error:', e);
+      if (available) {
+        window.setTimeout(() => {
+          if (slog.readyState !== EventSource.CLOSED) return;
+          console.log('slog: reconnecting...');
+          setupSlog();
+        }, 10000);
+      }
+    }
+
+    set(state => { state.slogstream = slog });
+  }, [sessions]);
+
+  const onInput = useCallback((ses: string, e: string) => {
+    const term = useTermState.getState().sessions[ses].term;
+    let belts: Array<Belt> = [];
+    let strap = '';
+
+    while (e.length > 0) {
+      let c = e.charCodeAt(0);
+
+      //  text input
+      //
+      if (c >= 32 && c !== 127) {
+        strap += e[0];
+        e = e.slice(1);
+        continue;
+      } else if ('' !== strap) {
+        belts.push({ txt: strap.split('') });
+        strap = '';
+      }
+
+      //  special keys/characters
+      if (0 === c) {
+        term.write('\x07');  //  bel
+      }
+      else if (8 === c || 127 === c) {
+        belts.push({ bac: null });
+      }
+      else if (13 === c) {
+        belts.push({ ret: null });
+      }
+      else if (c <= 26) {
+        belts.push({ mod: { mod: 'ctl', key: String.fromCharCode(96 + c) } });
+      }
+
+      //  escape sequences
+      //
+      if (27 === c) {  //  ESC
+        e = e.slice(1);
+        c = e.charCodeAt(0);
+        if (91 === c || 79 === c) {  //  [ or O
+          e = e.slice(1);
+          c = e.charCodeAt(0);
+          switch (c) {
+            case 65: belts.push({ aro: 'u' }); break;
+            case 66: belts.push({ aro: 'd' }); break;
+            case 67: belts.push({ aro: 'r' }); break;
+            case 68: belts.push({ aro: 'l' }); break;
+          //
+            case 77:
+              const m = e.charCodeAt(1) - 31;
+              if (1 === m) {
+                const c = e.charCodeAt(2) - 32;
+                const r = e.charCodeAt(3) - 32;
+                belts.push({ hit: { r: term.rows - r, c: c - 1 } });
+              }
+              e = e.slice(3);
+              break;
+          //
+            default: term.write('\x07'); break;  //  bel
+          }
+        }
+        else if (c >= 97 && c <= 122) {  //  a <= c <= z
+          belts.push({ mod: { mod: 'met', key: e[0] } });
+        }
+        else if (c === 46) {  //  .
+          belts.push({ mod: { mod: 'met', key: '.' } });
+        }
+        else if (c === 8 || c === 127) {
+          belts.push({ mod: { mod: 'met', key: { bac: null } } });
+        }
+        else {
+          term.write('\x07'); break;  //  bel
+        }
+      }
+
+      e = e.slice(1);
+    }
+    if ('' !== strap) {
+      belts.push({ txt: strap.split('') });
+      strap = '';
+    }
+    belts.map(b => {  //NOTE  passing api.term.sendBelt makes `this` undefined!
+      api.term.sendBelt(ses, b);
+    });
+  }, [sessions, api.term]);
```

Comment:
consider refactoring into `inputToBelts` (where the meat of the logic is and can be outside component scope) and then `onInput` would just call that and forward the results to the api.